File: tests/org/elixir_lang/reference/callable/Issue687Test.java
package org.elixir_lang.reference.callable;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.testFramework.fixtures.LightCodeInsightFixtureTestCase;
import org.elixir_lang.psi.ElixirIdentifier;
import org.elixir_lang.psi.UnqualifiedNoArgumentsCall;
import org.elixir_lang.psi.call.Call;

public class Issue687Test extends LightCodeInsightFixtureTestCase {
    /*
     * Tests
     */

    public void testRepeatedMapValueInMatch() {
        myFixture.configureByFiles("repeated_map_value_in_match.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);
        assertInstanceOf(elementAtCaret, LeafPsiElement.class);

        PsiElement parent = elementAtCaret.getParent();

        assertNotNull(parent);
        assertInstanceOf(parent, ElixirIdentifier.class);

        PsiElement grandParent = parent.getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        Call grandParentCall = (Call) grandParent;

        PsiReference reference = grandParentCall.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNotNull(resolved);

        assertInstanceOf(resolved, UnqualifiedNoArgumentsCall.class);
        assertEquals(resolved.getText(), "nine_id");

        assertEquals(resolved.getParent().getParent().getText(), "nine_id: nine_id");
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/issue_687";
    }
}

File: tests/org/elixir_lang/reference/callable/Issue687Test.kt
package org.elixir_lang.reference.callable

import org.elixir_lang.psi.ElixirIdentifier
import com.intellij.psi.impl.source.tree.LeafPsiElement
import org.elixir_lang.psi.UnqualifiedNoArgumentsCall
import org.elixir_lang.PlatformTestCase
import org.elixir_lang.psi.call.Call

class Issue687Test : PlatformTestCase() {
    fun testRepeatedMapValueInMatch() {
        myFixture.configureByFiles("repeated_map_value_in_match.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)
        assertInstanceOf(elementAtCaret, LeafPsiElement::class.java)

        val parent = elementAtCaret!!.parent
        assertNotNull(parent)
        assertInstanceOf(parent, ElixirIdentifier::class.java)

        val grandParent = parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)
        val grandParentCall = grandParent as Call

        val reference = grandParentCall.reference
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNotNull(resolved)
        assertInstanceOf(resolved, UnqualifiedNoArgumentsCall::class.java)
        assertEquals(resolved!!.text, "nine_id")
        assertEquals(resolved.parent.parent.text, "nine_id: nine_id")
    }

    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/issue_687"
}
